File: Sales_P3/script/pricebudget.py
import numpy as np


# Define a custom function to calculate the budget price
def bud_price(df, col_name):
    col = col_name
    df = (
        df.loc[:, ["Version", "Month", "Profit Ctr", col, "Qty", "Sales_LC"]]
        .query('Version == "Budget"')
        .groupby(["Version", "Profit Ctr", col])
        .agg(Qty=("Qty", "sum"), Sales_LC=("Sales_LC", "sum"))
        .assign(bud_price=lambda x: np.round(x.Sales_LC / x.Qty, 0))
        .reset_index()
        .dropna(subset=[col])
        .loc[:, ["Profit Ctr", col, "bud_price"]]
    )
    df["mapping key"] = df[col]
    return df


# 3 different groups
def df_div_e(df):
    return df.query('Division == "E"')


def df_div_p(df):
    return df.query('`Outlet name`.isin(["PL EAC", "PL HYD", "PL MES"])')


def df_pl_cm(df):
    return df.query('`Outlet name`.isin(["PL CM CCN", "PL CM CVS", "PL CM PSS"])')


# Create budget price tables for 3 different groups
def bud_price_div_e(df):
    df = df_div_e(df)
    df = bud_price(df, "Customer Material")
    return df


def bud_price_div_p(df):
    df = df_div_p(df)
    df = bud_price(df, "Product")
    return df


def bud_price_pl_cm(df):
    df = df_pl_cm(df)
    df = bud_price(df, "CM Cluster")
    return df


# mapping between between actual and budget
def spv_mapping(df, bud_price, col_name):
    return df.merge(bud_price, on=["Profit Ctr", col_name], how="left")

File: Sales_P3/src/dplyr/price_budget.R
library(dplyr)


# Define a custom function to calculate the budget price ----
bud_price <- function(df, col_name) {
  col <- col_name
  df <- df |>
    select(c(
      "Version",
      "Month",
      "Profit Ctr",
      col,
      "Qty",
      "Sales_LC"
    )) |>
    filter(df$Version == "Budget") |>
    group_by(pick(
      "Version",
      "Profit Ctr",
      col,
    )) |> # "Month" is excluded
    summarise(
      Qty = sum(df$Qty),
      Sales_LC = sum(df$Sales_LC),
      bud_price = round(df$Sales_LC / df$Qty, 0),
    ) |>
    ungroup() |>
    tidyr::drop_na() |>
    select(!c("Version", "Qty", "Sales_LC")) |>
    mutate(`mapping key` = df[col], .before = "bud_price")

  return(df)
}

# 3 different groups ----
df_div_e <- function(df) {
  df <- df |>
    filter(df$Division == "E")
  return(df)
}

df_div_p <- function(df) {
  df <- df |>
    filter(df["Outlet name"] %in% c("PL EAC", "PL HYD", "PL MES"))
  return(df)
}

df_pl_cm <- function(df) {
  df <- df |>
    filter(df["Outlet name"] %in% c("PL CM CCN", "PL CM CVS", "PL CM PSS"))
  return(df)
}

# Create budget price tables for 3 different groups ----
bud_price_div_e <- function(df) {
  df <- df |>
    df_div_e() |>
    bud_price("Customer Material")
  return(df)
}

bud_price_div_p <- function(df) {
  df <- df |>
    df_div_p() |>
    bud_price("Product")
  return(df)
}

bud_price_pl_cm <- function(df) {
  df <- df |>
    df_pl_cm() |>
    bud_price("CM Cluster")
  return(df)
}

# mapping between between actual and budget ----
spv_mapping <- function(df, bud_price, col_name) {
  df <- df |>
    left_join(bud_price, by = c("Profit Ctr", col_name))
  return(df)
}
